client/mm/exchange_adaptor.go

```diff
@@ -29,40 +31,61 @@ type botBalance struct {
 	Cancel(oidB dex.Bytes) error
-	MaxBuy(host string, base, quote uint32, rate uint64) (*core.MaxOrderEstimate, error)
-	MaxSell(host string, base, quote uint32) (*core.MaxOrderEstimate, error)
-	DEXBalance(assetID uint32) (*botBalance, error)
-	MultiTrade(pw []byte, form *core.MultiTradeForm) ([]*core.Order, error)
-	MaxFundingFees(fromAsset uint32, host string, numTrades uint32, fromSettings map[string]string) (uint64, error)
-	FiatConversionRates() map[uint32]float64
+	DEXTrade(rate, qty uint64, sell bool) (*core.Order, error)
 	ExchangeMarket(host string, base, quote uint32) (*core.Market, error)
+	MultiTrade(placements []*multiTradePlacement, sell bool, driftTolerance float64, currEpoch uint64, dexReserves, cexReserves map[uint32]uint64) []*order.OrderID
+	CancelAllOrders() bool
+	ExchangeRateFromFiatSources() uint64
+	OrderFeesInUnits(sell, base bool, rate uint64) (uint64, error)
+	SubscribeOrderUpdates() (updates <-chan *core.Order)
+	SufficientBalanceForDEXTrade(rate, qty uint64, sell bool) (bool, error)
 }
 
-// botCexAdaptor is an interface used by bots to access functionality
-// related to a CEX. Some of the functions are implemented by libxc.CEX, but
-// have some differences, since this interface is meant to be used by only
-// one caller. For example, Trade does not take a subscriptionID, and
-// SubscribeTradeUpdates does not return one. Deposit is not available
-// on libxc.CEX as it involves sending funds from the DEX wallet, but it is
-// exposed here.
+// botCexAdaptor is an interface used by bots to access CEX related
+// functions. Common functionality used by multiple market making
+// strategies is implemented here. The functions in this interface
+// take assetID parameters, unlike botCoreAdaptor, to support a
+// multi-hop strategy.
 type botCexAdaptor interface {
-	CEXBalance(assetID uint32) (*botBalance, error)
 	CancelTrade(ctx context.Context, baseID, quoteID uint32, tradeID string) error
 	SubscribeMarket(ctx context.Context, baseID, quoteID uint32) error
 	SubscribeTradeUpdates() (updates <-chan *libxc.Trade, unsubscribe func())
-	Trade(ctx context.Context, baseID, quoteID uint32, sell bool, rate, qty uint64) (*libxc.Trade, error)
+	CEXTrade(ctx context.Context, baseID, quoteID uint32, sell bool, rate, qty uint64) (*libxc.Trade, error)
+	SufficientBalanceForCEXTrade(baseID, quoteID uint32, sell bool, rate, qty uint64) (bool, error)
 	VWAP(baseID, quoteID uint32, sell bool, qty uint64) (vwap, extrema uint64, filled bool, err error)
-	Deposit(ctx context.Context, assetID uint32, amount uint64, onConfirm func()) error
-	Withdraw(ctx context.Context, assetID uint32, amount uint64, onConfirm func()) error
+	PrepareRebalance(ctx context.Context, assetID uint32) (rebalance int64, dexReserves, cexReserves uint64)
+	FreeUpFunds(assetID uint32, cex bool, amt uint64, currEpoch uint64)
+	Deposit(ctx context.Context, assetID uint32, amount uint64) error
+	Withdraw(ctx context.Context, assetID uint32, amount uint64) error
 }
```

Comment:
These intermediate interfaces seem superfluous to me. Examining the way that the test structures (`tCore`, `tBotCoreAdaptor`, `tCEX`, and `tBotCexAdaptor`) are designed and used seems to confirm my feeling here. So I took a swing at eliminating these intermediate interfaces and just using `unifiedExchangeAdaptor`. We already have `libxc.CEX` and `clientCore` for testing stubs anyway. I think it's an improvement.

https://github.com/buck54321/dcrdex/compare/5f9422e...buck54321:dcrdex:nomiddleinterfaces

I didn't rework all the tests there, but I did do the test structures in mm_test.go and one test (`TestBasisPrice`) just as part of the proof-of-concept.

Reply:
I understand that what I'm asking here is challenging work to deeply reconfigure the tests. My view is that your refactor **is so righteous**, that we can literally eliminate an entire layer of abstraction. IDE navigation will be so much better. Now seems like the right time to adapt at this level and scope.

Reply:
I'll look into it, but testing the bot logic is so much simpler with these interfaces. Take for example `OrderFeesInUnits`, the rebalancing functions, and even `MultiTrade`. These have pretty complicated logic. By having the interfaces, it's way easier to directly test the bots. `TestBasisPrice` may have been easy to update, but the arb+mm and arb rebalancing tests will need to have so much more additional logic that does not involve directly testing the bots. Why is this better?